# Coding Temple's Data Analytics Program:
---
## Python II: Map, Filter, Reduce, Lambda & Recursion
---

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [ ]:
# lambda >input< x: what we want to do
# Let's take a look at how we can create a function to add two to a variable
def add_two(x):
    return x+2

# Then we can test it:
print(add_two(4))
# This is great if we are looking to reuse this function over and over again.
# But what if this was something I only wanted to do once?
# Lambda comes into play!
a_var = 4
print((lambda x: x+2)(a_var))

6
6


In terms of time and space complexity, they are the same in performance as a normal function, if not worse at times.

Again, Lambda should be used when we want to do a single function and not reuse it.

#### Saving to a Variable

In [4]:
# We can save a lambda function to a variable for reusability
# Do not recommend this as you can write a function to do the same thing.
f_test = lambda x: x+2
f_test(4)

6

In [5]:
f_test(78)

80

#### Multiple Inputs

In [7]:
# No variable
print((lambda x,y,z: x * y * z)(3,5,8))

# If we wanted to save the lambda function as a variable, we could as well with multiple inputs:
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [9]:
def multiply(function, num):
    """
    f expects to see a function (lambda function) 
    num expects a single number
    """
    return function(num)

multiply(lambda x: x*x, 4)

# Not only can we pass lambda functions into another function, but we pass actual functions in as well.
def squared_nums(num):
    return num ** 2

multiply(squared_nums, 4)

16

#### Returning a Lambda from a Function

In [5]:
def multiply_test(num):
    return num * 4

# Let's try function inside of another function
def return_func():

    def multiply(num):
        return num*2
    return multiply

f_return = return_func()
print(f_return(4))

#print(return_func(4)) #need to list num as an arg IF you do this
# What if I wanted to instead return a lambda function from my regular function?
def return_lamb(b,c):
    return lambda x,a: x+a+b+c

r_lamb = return_lamb(4,5)
print(r_lamb)
print(r_lamb(5,9))

8
<function return_lamb.<locals>.<lambda> at 0x00000215469239A0>
23


#### If Statements within Lambdas

In [21]:
# Lambdas are set up to be able to take in conditional statements
# Lambda x: True if (condition happens) else False
# Lambda x: do something if (condition happens) else do something else
f_conditional = lambda num: num * 2 if num > 10 else num + 2
print(f_conditional(8))
print(f_conditional(20))

10
40


In [7]:
lst = [1,2,3,4,5,6,7,8,9,10,11,12,13]

[f_conditional(num) for num in lst]

NameError: name 'f_conditional' is not defined

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [6]:
# Write out a function that cubes any number
f= lambda x: x**3

print(f(2))

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [9]:
numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]

# Normally, in order to use a function with this, we would iterate through the list!
%time
def squared_normal(val):
    return val **2
lst = []
for num in numbers:
    lst.append(squared_normal(num))

print(lst)

CPU times: total: 0 ns
Wall time: 0 ns
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1, 1936, 1]


In [16]:
# By using a map function, we can iterate through a list and apply a function without ever having to iterate ourselves.

"""
map(function, iterable)
An iterable is something we can loop over; such as a list, dict, tuple, set, or a string.
Normally, you would use map with a predefined function, but we can use it with a lambda statement.
"""
%time
def squared_map(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2 **2
    else:
        return num, num2

numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]
more_nums = [4,10,3,2,6,10,15,3,4,7,6,0,41,2]

# In order to use the squared_map function without a map, we would need to iterate through the lists
# To do so effectively, we would need to utilize the zip() function, which brings together two iterables
# And returns each value in those iterables.
#for val1, val2 in zip(numbers, more_nums):
 #   print(squared_map(val1, val2))

''' 
When we use map against two or more iterable objects, we get returned a list of tuples
Each value in the tuple represents a value in the corresponding input list
Note: When we use map, a map object is what is returned. To make sure that we are appropriately 
Displaying the information, we use the list function
'''
squared_numbers_mapped = list(map(squared_map, numbers, more_nums))
print(squared_numbers_mapped)

# When we use map against a single iterable object, we will be returned a list of values
# Where each value corresponds to the value in the input object.
another_squared_nums_mapped = list(map(squared_normal, numbers))
print(another_squared_nums_mapped)

CPU times: total: 0 ns
Wall time: 0 ns
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6), (20, 10), (11, 15), (4, 9), (9, 16), (36, 49), (25, 36), (1, 0), (44, 41), (1, 4)]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1, 1936, 1]


#### Using Lambda's with Map

In [43]:
"""
Instead of having to build out a function that we plan to use one time,
We can use a lambda inside the map function, allowing us to do everything in a single line of code

Syntax for this looks like:
map(lambda x: do something, iterable)
Once more, using lambda in map almost ALWAYS takes place in a single line of code!
"""

# Remember that we need to use the list function in conjunction with the map function in order to return the values given
# Otherwise, we return the map object, which is useless to us
list(map(lambda x,y: (x**2, y**2) if x < 10 and y < 10 else (x,y), numbers, more_nums))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (4, 9),
 (9, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 41),
 (1, 4)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [26]:

print(list(map(lambda x,y: (x*2 -1, y*2 -1), numbers, more_nums)))

# Let's reflect back to today's whiteboard:
# How could I use a map function to add the asterisks to the string values using a lambda?
picture = [ '','abc', 'ded', '']
# Using string values inside the lambda function, and applying acrossed a list of strings
list(map(lambda x: '*' + x + '*' if len(x) > 1 else '*****', picture))

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11), (39, 19), (21, 29), (3, 5), (5, 7), (11, 13), (9, 11), (1, -1), (87, 81), (1, 3)]


['*****', '*abc*', '*ded*', '*****']

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [51]:
# Syntax for this looks like:
# list(filter(function, iterable(s)))
# A lot like a mask
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angela']

def a_names(name):
    # In this function, we will check to see if the first letter of each name begins with 'A'
    if name[0].lower() == 'a':
        return True
    else: # If it is not an 'a'
        return False
    
# Our function returns a boolean value. It will allow our filter() function to return only the names in a list that start with "A"
list(filter(a_names, names))

['Andy', 'Angela']

In [68]:
# A string is an iterable as well as indexable. So when index the zeroith position of a string value,
# we return the very first value in that string
for name in names:
    print(name, name[0], name[1:])

Bob B ob
Andy A ndy
Max M ax
Evan E van
Angela A ngela


#### Using Lambda's with Filter()

In [72]:
# We can also use Lambdas with the filter function! 
# This is great if you are looking to filter data with a single-use function instead of having to define
# And create another function like we did above

# When a lambda, or comprehension, we typicall write out the value we want returned FIRST, then follow up with the if
# The best way to think about this is just a reverse loop or reversed if statement
list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

# Instead of this:
for name in names:
    if name[0].lower() == 'a':
        print(True)
    else:
        print(False)

# We say 
for name in names:
    lambda name: True if name[0].lower == 'a' else False
    
# List comprehension method:
[True if name[0].lower() == 'a' else False for name in names]

['Andy', 'Angela']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [35]:
import statistics
import numpy as np
numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]
# Using the statstics module to gather the mean
print(statistics.mean(numbers))

# Using the numpy library to gather the mean as well
print(np.mean(numbers))

# There are three ways to go about this.

# Number 1: Using vanilla Python
print(list(filter(lambda x: True if (sum(numbers)/len(numbers))<x else False,numbers)))

# Number 2: Using the numpy library
print(list(filter(lambda x: True if (np.mean(numbers))<x else False,numbers)))

# Number 3: Using the Statistics module
print(list(filter(lambda x: True if (statistics.mean(numbers))<x else False,numbers)))

10.428571428571429
10.428571428571429
[11, 20, 15, 20, 11, 44]
[11, 20, 15, 20, 11, 44]
[11, 20, 15, 20, 11, 44]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Show a way to break the program by creating an infinite loop!

In [81]:
# Start off with defining a function!
def add_nums_breakable(num):
    # Without a base case, we will end up infinitely looping through our program. Luckily, VS code has built in safeguards 
    # For you!
    if num>-20000:
        return num + add_nums_breakable(num-1)
    else:
        return num

add_nums_breakable(5)

RecursionError: maximum recursion depth exceeded

#### Implementing a Base Case

In [86]:
'''
What is a base case?
A base case is what cause the function to break out of it's infinite loop!
It evaluates the value of the number or string at the moment it recurrs. To stop from going any further.
These need to be implemented at the BEGINNING of your function, not after the recursive call.
'''

def add_nums(num):
    # Implement a base case
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
    return num + add_nums(num-1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [90]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    # Base case to begin with!
    if num <= 1:
        print(f'factorial({num}) = {num}')
        return 1
    else:
        print(f'factorial({num}) = {num} * factorial({num - 1})')
        return num * factorial(num -1) 

print(factorial(5))
5*4*3*2*1

factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
factorial(1) = 1
120


120

#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [51]:
# How can we use recursion to subtract two numbers from each other?
# To start with, we will define a function
# should be thought of as num1-num2


'''my work'''

def two_nums(x,y):
    if y==0:
        return x     
    else:
        x -= 1
        y -= 1
        return two_nums(x,y)

print(two_nums(13,10))






























3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [103]:
# Built in Python Generator
print([i for i in range(10)])

# Generator package up the values and yield them as opposed to returning them. Then, we can iterate through that generator
# Object to return each value packaged inside of it.
def my_range(stop, start = 0, step = 1):
    while start < stop:
        yield start # Yield means that is going to be a generator. It will not return anything when the function is run
        start += step
print([i for i in my_range(10)])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [55]:
def squaretill0(num):
    while num>0:
        yield num**2
        print(num**2)
        num-=1

print([i for i in squaretill0(5)])


25
16
9
4
1
[25, 16, 9, 4, 1]
